### PR TITLE
CC-1126: Add C# support to grep challenge.

### PR DESCRIPTION
**Pull request by ryan-gang**

### CODE REVIEW

File: compiled_starters/csharp/src/Program.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
+using System;
+using System.IO;
+
+// You can use print statements as follows for debugging, they'll be visible when running tests.
+Console.WriteLine("Logs from your program will appear here!");
+
+// Uncomment this block to pass the first stage
```
Comment (rohitpaulk) [inline compiled_starters/csharp/src/Program.cs:7]: @ryan-gang can we make the uncommented section shorter? This appears in the UI, so ideally it'd be a small diff, like the Python one:  

Screenshot 2024-04-09 at 16 19 02

File: compiled_starters/csharp/src/Program.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+// string inputLine = Console.In.ReadToEnd();
+// if (MatchPattern(inputLine, pattern))
+// {
+//     Environment.Exit(0);
+// }
+// else
+// {
+//     Environment.Exit(1);
+// }
+// static bool MatchPattern(string inputLine, string pattern)
```
[the remark this answers is not in the document]
Comment (rohitpaulk) [reply compiled_starters/csharp/src/Program.cs:29]: Let's find a C# formatter / linter and try to run it on this - we want code to always look well-formatted

**Review by rohitpaulk — CHANGES_REQUESTED** [review]
Notes added

File: compiled_starters/csharp/src/Program.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+// Uncomment this block to pass the first stage
+//
+// if (args[0] != "-E") {
+//     Console.WriteLine("Expected first argument to be '-E'");
+//     Environment.Exit(2);
+// }
+// string pattern = args[1];
+// string inputLine = Console.In.ReadToEnd();
+// if (MatchPattern(inputLine, pattern)) {
```
Comment (rohitpaulk) [inline compiled_starters/csharp/src/Program.cs:23]: @ryan-gang let's make this even smaller - can we do just the last 5 lines? That'd be similar to what we do for Python

File: solutions/csharp/01-init/code/src/Program.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    } else {
+        throw new ArgumentException($"Unhandled pattern: {pattern}");
+    }
+}
+
+if (args[0] != "-E") {
+    Console.WriteLine("Expected first argument to be '-E'");
+    Environment.Exit(2);
+}
+string pattern = args[1];
```
Comment (rohitpaulk) [inline solutions/csharp/01-init/code/src/Program.cs:16]: @ryan-gang can you confirm how we reviewed formatting here? How do we know that this is C# code that will look "well-formatted" to a senior dev?

**Comment by ryan-gang** [comment]
Reduced the commented section size further.
Formatted code using `dotnet format`, and added white spaces between blocks.

**Review by rohitpaulk — APPROVED** [review]

**Pull request merged**